src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+
+-- | Data type for root of sized merkle tree.
+data SizedMerkleRoot a = SizedMerkleRoot
+    { smrHash :: Hash LBS.ByteString  -- ^ returns root 'Hash' of Merkle Tree
+    , smrSize :: Word8 -- ^ size of root node,
+                       -- size is defined as number of leafs in this subtree
+    } deriving (Show, Eq, Ord, Generic, ByteArrayAccess , Serialise, Typeable)
+
+
+data SizedMerkleTree a
+  = SizedMerkleEmpty
+  | SizedMerkleTree !(SizedMerkleNode a)
+  deriving (Eq, Generic)
+
+
+instance Foldable SizedMerkleTree where
+    foldMap _ SizedMerkleEmpty      = mempty
+    foldMap f (SizedMerkleTree n) = foldMap f n
+
+    null SizedMerkleEmpty = True
+    null _                = False
+
+    length SizedMerkleEmpty      = 0
+    length (SizedMerkleTree n) = fromInteger (toInteger (smrSize (mRoot n)))
+
+instance Show a => Show (SizedMerkleTree a) where
+  show SizedMerkleEmpty = "Empty Sized Merkle tree "
+  show (SizedMerkleTree n) = "Sized Merkle tree:  " <> Prelude.show n
+  --show tree = "Merkle tree: "  <> Prelude.show (toList tree)
+
+type LeafIndex = Int
+
+data SizedMerkleNode a
+    = SizedMerkleBranch { mRoot  :: !(SizedMerkleRoot a)
+                        , mLeft  :: !(SizedMerkleNode a)
+                        , mRight :: !(SizedMerkleNode a)}
+    | SizedMerkleLeaf { mRoot  :: !(SizedMerkleRoot a)
+                      , mIndex :: !LeafIndex
+                      , mVal   :: !a}
+    deriving (Eq, Show, Generic)
+
+instance Foldable SizedMerkleNode where
+  foldMap f x = case x of
+    SizedMerkleLeaf {mVal} -> f mVal
+    SizedMerkleBranch {mLeft, mRight} -> foldMap f mLeft `mappend` foldMap f mRight
+
+toLazyByteString :: Builder -> LBS.ByteString
+toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
+
+mkLeaf :: HasHash a => (LeafIndex, a) -> SizedMerkleNode a
+mkLeaf (i, a) =
+    SizedMerkleLeaf
+    { mVal  = a
+    , mIndex = i
+    , mRoot = SizedMerkleRoot (unsafeHash a) -- unsafeHash since we need to hash to ByteString
+                              1 -- size of leaf node is 1
+    }
+
+mkBranch :: SizedMerkleNode a -> SizedMerkleNode a -> SizedMerkleNode a
+mkBranch l r =
+    SizedMerkleBranch
+    { mLeft  = l
+    , mRight = r
+    , mRoot  = mkBranchRootHash (mRoot l) (mRoot r)
+    }
+
+mkBranchRootHash :: SizedMerkleRoot a -- ^ left merkle root
+                 -> SizedMerkleRoot a -- ^ right merkle root
+                 -> SizedMerkleRoot a
+mkBranchRootHash (SizedMerkleRoot (AbstractHash hl) sl)
+                 (SizedMerkleRoot (AbstractHash hr) sr)
+  = SizedMerkleRoot
+     (hash $ toLazyByteString $ mconcat
+        [ byteString (one sl)
+        , byteString (one sr)
+        , byteString (convert hl)
+        , byteString (convert hr) ])
+     (sl + sr)
+
+-- | Smart constructor for 'SizedMerkleTree'.
+mkSizedMerkleTree :: (HasHash a, Foldable t) => t a -> SizedMerkleTree a
+mkSizedMerkleTree x = mkSizedMerkleTree' (zip [0,1..] (toList x))
```

Comment:
Why not `fromFoldable`?

Reply:
I'm not aware of `fromFoldable`, will look it up.

Reply:
Do you mean I should use some other function to implement this? Or is it just a comment on changing the function name to fromFoldable?

Reply:
I believe this is about naming: https://www.stackage.org/lts-11.10/hoogle?q=fromFoldable